File: qc-core/src/main/java/com/eager/questioncloud/core/domain/user/implement/UserRegister.java
package com.eager.questioncloud.core.domain.user.implement;

import com.eager.questioncloud.core.domain.user.model.User;
import com.eager.questioncloud.core.domain.user.repository.UserRepository;
import com.eager.questioncloud.core.domain.user.vo.AccountType;
import com.eager.questioncloud.core.exception.CustomException;
import com.eager.questioncloud.core.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class UserRegister {
    private final UserRepository userRepository;

    public User register(User user) {
        if (checkDuplicateEmail(user.getUserInformation().getEmail())) {
            throw new CustomException(Error.DUPLICATE_EMAIL);
        }
        if (checkDuplicatePhone(user.getUserInformation().getPhone())) {
            throw new CustomException(Error.DUPLICATE_PHONE);
        }
        if (checkDuplicateSocialUidAndAccountType(
            user.getUserAccountInformation().getSocialUid(),
            user.getUserAccountInformation().getAccountType())) {
            throw new CustomException(Error.DUPLICATE_SOCIAL_UID);
        }
        return userRepository.save(user);
    }

    public Boolean checkDuplicatePhone(String phone) {
        return userRepository.checkDuplicatePhone(phone);
    }

    public Boolean checkDuplicateEmail(String email) {
        return userRepository.checkDuplicateEmail(email);
    }

    public Boolean checkDuplicateSocialUidAndAccountType(String socialUid, AccountType accountType) {
        return userRepository.checkDuplicateSocialUidAndAccountType(socialUid, accountType);
    }
}

File: qc-api/qc-user-api/src/main/kotlin/com/eager/questioncloud/user/register/implement/UserRegister.kt
package com.eager.questioncloud.user.register.implement

import com.eager.questioncloud.point.implement.UserPointManager
import com.eager.questioncloud.social.SocialAPIManager
import com.eager.questioncloud.social.SocialPlatform
import com.eager.questioncloud.user.domain.User
import com.eager.questioncloud.user.domain.UserAccountInformation
import com.eager.questioncloud.user.domain.UserInformation
import com.eager.questioncloud.user.dto.CreateUser
import com.eager.questioncloud.user.enums.AccountType
import com.eager.questioncloud.user.enums.UserStatus
import com.eager.questioncloud.user.enums.UserType
import com.eager.questioncloud.user.infrastructure.repository.UserRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class UserRegister(
    private val userRegistrationValidator: UserRegistrationValidator,
    private val userPointManager: UserPointManager, // TODO 임시 의존성
    private val socialAPIManager: SocialAPIManager,
    private val userRepository: UserRepository,
) {
    @Transactional
    fun create(createUser: CreateUser): User {
        val userAccountInformation = createUserAccountInformation(createUser)
        val userInformation = UserInformation.create(createUser)
        userRegistrationValidator.validate(userAccountInformation, userInformation)

        val user = userRepository.save(
            User.create(
                userAccountInformation,
                userInformation,
                UserType.NormalUser,
                UserStatus.PendingEmailVerification
            )
        )

        userPointManager.init(user.uid)
        return user
    }

    private fun createUserAccountInformation(createUser: CreateUser): UserAccountInformation {
        if (createUser.accountType == AccountType.EMAIL) {
            return UserAccountInformation.createEmailAccountInformation(createUser.password!!)
        }

        val socialUid = socialAPIManager.getSocialUid(
            createUser.socialRegisterToken!!,
            SocialPlatform.valueOf(createUser.accountType.value)
        )

        return UserAccountInformation.createSocialAccountInformation(socialUid, createUser.accountType)
    }
}
